src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,193 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
+    public MyLinkedList() {
+    }
+
     @Override
     public void add(T value) {
+        Node<T> oldLast = last;
+        Node<T> newNode = new Node<>(oldLast, value, null);
+        last = newNode;
+        if (oldLast != null) {
+            oldLast.next = newNode;
+        } else {
+            first = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkPositionIndex(index);
+        if (index == size) {
+            add(value);
+        } else {
+            if (index == 0) {
+                insertToBeginning(value);
+            } else {
+                insertBeforeIndex(value, index);
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (!list.isEmpty()) {
+            for (T element : list) {
+                add(element);
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> node = findNodeByIndex(index);
+        return node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkPositionIndex(index);
+        Node<T> node = findNodeByIndex(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> node = findNodeByIndex(index);
+        T removingValue = node.value;
+        unlink(node);
+        size--;
+        return removingValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = first;
+        while (currentNode != null) {
+            if (equalsElement(currentNode.value, object)) {
+                unlink(currentNode);
+                size--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private void checkPositionIndex(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Can't add value in position: " + index
+                    + ". Index is invalid.");
+        }
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        checkElementIndex(index);
+        Node<T> currentNode;
+        if (index < (size << 1)) {
```

Comment:
The condition `index < (size << 1)` seems incorrect for determining whether to iterate from the start or the end of the list. It should be `index < (size >> 1)` to iterate only half of the list, improving performance.